server/db/mgo.go

```diff
@@ -713,3 +716,100 @@ func (m *mongoDB) GetAddressInfo(ctx context.Context, hash string) (*types.Addre
 }
 
 //endregion Address
+
+// start region Proposal
+
+func (m *mongoDB) AddVoteToProposal(ctx context.Context, proposalInfo *types.ProposalDetail, voteOption uint64) error {
+	currentProposal, err := m.ProposalInfo(ctx, proposalInfo.ID)
+	if err != nil {
+		currentProposal = proposalInfo
+	}
+	// update number of vote choices
+	switch voteOption {
+	case 0:
+		proposalInfo.NumberOfVoteAbstain = currentProposal.NumberOfVoteAbstain + 1
+	case 1:
+		proposalInfo.NumberOfVoteYes = currentProposal.NumberOfVoteYes + 1
+	case 2:
+		proposalInfo.NumberOfVoteNo = currentProposal.NumberOfVoteNo + 1
+	}
+	proposalInfo.UpdateTime = time.Now().Unix()
+	if err := m.upsertProposal(proposalInfo); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (m *mongoDB) UpsertProposal(ctx context.Context, proposalInfo *types.ProposalDetail) error {
+	currentProposal, err := m.ProposalInfo(ctx, proposalInfo.ID)
+	if err == nil { // need to update these stats from db first
+		proposalInfo.NumberOfVoteAbstain = currentProposal.NumberOfVoteAbstain
+		proposalInfo.NumberOfVoteYes = currentProposal.NumberOfVoteYes
+		proposalInfo.NumberOfVoteNo = currentProposal.NumberOfVoteNo
+	}
+	proposalInfo.UpdateTime = time.Now().Unix()
+	if err := m.upsertProposal(proposalInfo); err != nil {
+		return err
+	}
+	return nil
```

Comment:
Why we need another function here ?

Reply:
It is used in two function so I separated it as a private one. It might have a naming convention problem either.